### ISSUE TITLE
[Bug]: Colons ':' in urls are not passed through to Story from args

### ISSUE THREAD

**Issue by jasonmurphy7** [storybookjs/storybook#21032]
### Describe the bug

When using the "open canvas in new tab" option and using the args parameter to pass values, there is an issue with strings containing a colon not being passed through.

For example:

?args=titleIcon:test
will work fine

?args=titleIcon:actions:info
will fail to pass titleIcon as it contains a ':'

Presuming because the JSON is having issues determining where the key/values starts and ends.

Steps:
- open canvas in new tab
- edit url to contain "args=[key]:[some value with a colon]
- story will not receive the value.

**Comment by BikashSah999** [comment]
@jasonmurphy7  It is mentioned in the doc [here](https://storybook.js.org/docs/react/writing-stories/args#setting-args-through-the-url) that `values provided in the URL are limited to alphanumeric characters, spaces, underscores, and dashes. Any other types will be ignored and removed from the URL `

**Comment by jasonmurphy7** [comment]
Thanks for linking that @BikashSah999 - I'm guessing there isn't a way around this as it's simply by design?
If so this can likely by closed 🙂

**Comment by shilman** [comment]
This is by design in the current version, but we'd like to address it in the next iteration, so I'll leave this open for now.

**Comment by jasonmurphy7** [comment]
Awesome thanks @shilman

**Comment by BikashSah999** [comment]
Hi @shilman , Could I work in this issue then?

**Comment by shilman** [comment]
@BikashSah999 I think we need to design the serialization format for controls. This will be an 8.0 change.

**Comment by kasperpeulen** [comment]
@shilman I would like to work on it. See if we can find a balance between readability and parsability at the same time.

**Comment by TomaszSz368** [comment]
This bug is still present any chance it will be fixed soon?

**Comment by Sidnioulz** [comment]
@hdodov proposed to use Base 64 encoding for args with special characters in #33853. I'm not sure if this would entirely satisfy our requirements against XSS attacks (need to think about it). For sure, we'd only want to encode/decode serialisable content. We'd want to continue to support plain text when possible to help third-parties form valid URLs too.

@kasperpeulen what did you have in mind for this? Did you want to rely on URL encoding by any chance? I think this should theoretically provide the same flexibility benefits whilst remaining more compatible with the existing code and easier for third-parties to craft URLs.

The same security concerns remain either way. Can we think of a scenario where our users will parse/eval an arg's content and absolutely need to see upfront if the arg passed in the URL could be a malicious payload?

**Comment by yandodov** [comment]
@Sidnioulz 

> I'm not sure if this would entirely satisfy our requirements against XSS attacks (need to think about it).

Yes, Base 64 is not a security measure. It **encodes** data, which is different from **encryption** (used for security). Encoding can be easily reversed (which is what we want in this case), but encryption is not.

In any case, to prevent XSS, you just have to sanitize the input:
1. Encode it with Base 64 (so URL-unsafe characters can be transported)
2. Add the Base 64 string to the URL
3. User clicks on the URL
4. Decode the Base 64 string
5. **Sanitize it** by e.g. removing `<script>` tags (this is what prevents XSS)

But I'd argue that this is unnecessary. React _already_ protects against XSS. If you pass `<script>...</script>`, React would render it as-is (as a string). It wouldn't _actually_ render a script tag and introduce a vulnerability.

To introduce an actual vulnerability, the user-defined story would have to use [`dangerouslySetInnerHTML`](https://react.dev/reference/react-dom/components/common#dangerously-setting-the-inner-html), and React already makes this mistake difficult to commit by including **"dangerously"** in the property name.

In other words, I don't think that's Storybook's concern. The job of Storybook is to transport the data from the controls as-is via the URL. Then, it's a responsibility **of the user** to make sure that `dangerouslySetInnerHTML` is not misused in the components that consume said data.

**Edit:** I believe it should be like AWS's [Shared Responsibility Model](https://aws.amazon.com/compliance/shared-responsibility-model/):

- **Storybook** is responsible for the security of everything _surrounding_ the story — the controls, the manager, etc. It makes sure that XSS is not introduced in _the process of transmitting_ the data (e.g. not doing dumb stuff like `eval`-ing during serialization).

- **The user** is responsible for the security of everything _within_ the story — the rendering itself, once said data has reached the component safely thanks to Storybook. It's up to _them_ to inject HTML responsibly (which is an extremely bad practice in general anyways).

**Comment by Sidnioulz** [comment]
@hdodov our current security stance is that we want to prevent XSS attacks at the URL query param layer. Our stance reflects the **defense in depth** principle: rather than leave the responsibility entirely on users to sanitise dangerous input, we want to ensure no dangerous input can be submitted to their published Storybook instances by third-parties.

If we did end up adopting base64 or any other encoding (including URL encoding), we'd still want to sanitise inputs and delete them when they match potentially unsafe content. We might allow content like ":" but not "<>" for instance. If we were to allow users to retrieve unsanitised content from URL query params, we'd probably make that behaviour opt-in behind our own config option with our own stern documentation explaining the relevant risks.

@shilman @kylegach, I'd like to hear your input here. Should we relax our position? If so, how would we allow users to opt into dangerous behaviour? What task list / definition of done could we communicate to @hdodov for them to bring their PR to completion?

**Comment by tetslee** [comment]
I agree with @yandodov that sanitizing the url args should not be storybook's concern. However to add another option how about adding `url.decode` and `url.encode` functions to controls to let people define their own level of sanitisation?

e.g.
```
const sanitizeUrlArg = (raw) => <whitelist regex>.test(raw) ? raw : undefined;

export default {
  ...
  argTypes: {
    value: {
      name: 'json value',
      control: {
        type: 'text',
      },
      url: {
        encode: sanitizeUrlArg,
        decode: sanitizeUrlArg,
      },
    },
    ...
  },
};
```
Then people can use btoa or whatever other mapping they want to make their urls prettier.

Additionally can add some convenience modes:
`url: true` => convenience for identity encode/decode (allows everything)
`url: false` => convenience for always undefined encode/decode (never use save/read from url)
`url: undefined` => current behaviour
